### ISSUE TITLE
Possible refresh token race condition

### ISSUE THREAD

**Issue by baojr** [workos/authkit-remix#7]
Hello! I think I'm experiencing the same issue described in workos/authkit-nextjs#28. Though I see that the change there was made in the API rather than in the authkit-nextjs project itself.

When my application makes a few authenticated requests at the same time with a token that needs to be refreshed, only the first call seems to consistently work, which results in the user getting logged out seemingly arbitrarily in development. I added some logging and confirmed that subsequent requests made within about 3 seconds with the same token to the refresh API are failing. Is there anything I can do in my app to reduce the likelihood of this happening?

Here are the relevant parts of my app's logs:
```
2024-07-10T20:40:22.262Z Trace: error verifying token JWTExpired: "exp" claim timestamp check failed
2024-07-10T20:40:22.266Z Trace: Session invalid. Attempting refresh <token-1> 
2024-07-10T20:40:22.534Z Trace: Refresh successful: <token-2> 
2024-07-10T20:40:25.964Z Trace: error verifying token JWTExpired: "exp" claim timestamp check failed
2024-07-10T20:40:25.965Z Trace: Session invalid. Attempting refresh <token-1> 
2024-07-10T20:40:26.262Z Error Description: Refresh token already exchanged.
Failed to refresh. Deleting cookie and redirecting. OauthException: Error: invalid_grant
```

**Comment by anthonyringoet** [comment]
Seeing the same issue, resulting in getting logged out constantly as cookie is cleared at this point.

[Should be fixed on workOS API side](https://github.com/workos/authkit-nextjs/issues/28#issuecomment-2154481795) as stated but still seeing this in both dev and prod tenants.

**Comment by PaulAsjes** [comment]
The grace period on the API side is currently set at 2 seconds, which I think sounds correct. There's probably more we could do to debounce those requests on the library side, but first can you share some more of your findings? Having a refresh request 3 seconds after one already succeeded feels like there's something more going on here with stale data.

In any case, I think it's a good idea to add a check to not delete the cookie and redirect if the refresh failed due to the token already being exchanged.

**Comment by anthonyringoet** [comment]
@PaulAsjes Just went through the code: I think [there's a bug](https://github.com/workos/authkit-remix/blob/edd45f9c4b227beb9bd52d4464b585bf193bcd5d/src/session.ts#L49) in the Remix SDK that makes it impossible to actually refresh a token?

`commitSession` works by [dumping it into a set cookie header](https://remix.run/docs/en/main/utils/sessions), not only awaiting. Eg. like [how it's used to destroy](https://github.com/workos/authkit-remix/blob/edd45f9c4b227beb9bd52d4464b585bf193bcd5d/src/session.ts#L59) the session a bit lower in the code.

What currently happens: refresh is successful (confirmed by logs) but is not persisted to the session cookie. Next page load, it uses the old token: boom, destroys session cookie and you're logged out. Explains why the first endpoint still works after coming back after the initial access token expiration.

**Comment by anthonyringoet** [comment]
If that assumption is correct, the workOS API is doing it's job perfectly fine 😄

**Comment by baojr** [comment]
I lowered the access token duration to 1 minute to test @anthonyringoet's findings in my development environment. It does indeed look like the session cookie isn't getting updated. I made a request that should have refreshed the token according to the debug logs and didn't see any `set-cookie` header on the corresponding response. I waited several seconds after everything had finished loading and made another HTTP request that caused my app to reuse the same stale refresh token.

**Comment by PaulAsjes** [comment]
Thanks both, turns out @anthonyringoet is exactly right and that the refresh token isn't successfully updated. Working on a fix now!

**Comment by PaulAsjes** [comment]
Thanks for help all, the fix has now been released as `0.3.0`. Please bear in mind that the new version has **breaking changes** to how the library now works. Check out the updated readme for new integration instructions.
